### PR TITLE
select: add label-key property (#12713)

### PR DESCRIPTION
**Pull request by nanomen**
Fix https://github.com/ElemeFE/element/issues/12713

Add in el-select a new property "label-key" which the same as value-key. This allow to set a label for tag element when we pass value as Object to select component.

Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide ([中文](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.zh-CN.md) | [English](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.en-US.md) | [Español](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.es.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer relative issues for you PR.

### CODE REVIEW

**Comment by nanomen** [comment]
Hello! I don't know what's happend (about test error "fullscreen singleton"). In my local machine all tests is passed image

**Comment by wacky6** [comment]
Why is this necessary?

**Comment by nanomen** [comment]
@wacky6 If I using `label` as property of object it is empty. Issue [#12713](https://github.com/ElemeFE/element/issues/12713) is fully describe the bug. Look, please.

**Comment by ziyoung** [comment]
@nanomen Could you please update doc for this attribute? I think you can also add a tip for this.

**Comment by nanomen** [comment]
@ziyoung sure! one moment!

**Comment by nanomen** [comment]
@ziyoung done! Add in this page `/examples/docs/en-US/select.md`

File: examples/docs/en-US/select.md [hunk 1]
```diff
@@ -665,6 +665,7 @@ If the binding value of Select is an object, make sure to assign `value-key` as
 | multiple | whether multiple-select is activated | boolean | — | false |
 | disabled | whether Select is disabled | boolean | — | false |
 | value-key | unique identity key name for value, required when value is an object | string | — | value |
+| label-key | unique identity key name for label, required when value is an object | string | — | label |
```
Comment (ziyoung) [inline examples/docs/en-US/select.md]: - I think this attribute should not be required. Only when the label doesn't display, the attribute is needed. You can add a tip for this.
- Just copy these change to Chinese and Spanish docs.

File: packages/select/src/select.vue [hunk 2]
```diff
@@ -533,8 +537,9 @@
           }
         }
         if (option) return option;
-        const label = (!isObject && !isNull)
-          ? value : '';
+        const label = (isObject && !isNull)
+          ? value[this.labelKey] ? value[this.labelKey] : ''
```
Comment (ziyoung) [inline packages/select/src/select.vue]:
```js
const label = (isObject && !isNull)
  ? (value[this.labelKey] || '')
  : (!isObject && !isNull) ? value : '';
```

**Comment by nanomen** [comment]
@ziyoung add description in docs and refactor conditional statement, thank you!

**Comment by wacky6** [comment]
My opinion is **strongly against** this feature. I have seen a similar issue some time ago. From my perspective, `label-key` is permitting developer sloppiness and laziness, which should never be tolerated in the first place.

A select without el-option should show nothing, even values are specified. It's *developer's* responsibility to add selected option into el-option list.

This is **analogous to native HTML select**. See https://jsfiddle.net/dphut543/, native select shows empty value if bounded value is not in the option list.

The correct way of fixing #12713 is https://jsfiddle.net/p0h327dw/. Basically, ensure the selected value is contained in option list.

---

Object values are meant to enable representation of complex values (for example, configuration preset / combination of values). They are not supposed to contain the label, which is rarely required for application logic (or submitted to the server).

The reason that `value-key` was implemented is to determine equality of different objects. Another way is ask developer to write a comparison function (like the one in `Array.sort`), which is laborious (and usually not required).

**Comment by wacky6** [comment]
@nanomen 

By saying "laziness", I mean developers omitting the step of "ensuring selected options are contained in the list of available options, and think about what labels should be displayed". In other words, logically correct (Some might consider this pedantic). This step could be as simple as merging two lists together.

I understand that it is tempting to show object values in some way (i.e. label-key). But it does not make sense logically. Suppose bounded values are integers, and options are empty. Current implementation displays the bounded value. (https://jsfiddle.net/3xyopdf0/).

This is acceptable, but logically incorrect. Select means choosing a subset of available options:

* Why would someone show internal values instead of user-friendly labels to user?
* Why would select display an option that is not permitted (declared by el-option)?
* How could someone select an option when there none is available?

---

In fact, I would rather el-select ignores all values that are not presented in option list. The current behavior is due to legacy implementation.

**Pull request closed without merging**